### PR TITLE
Support multiple data fidelity dimensions in MultiFidelityGP models

### PR DESCRIPTION
**Pull request by AlexanderMouton**
Support multiple data fidelity dimensions in MultiFidelityGP models

## Motivation

Both the SingleTaskMultiFidelityGP and the FixedNoiseMultiFidelityGP only allowed for one data-fidelity dimension. This PR generalises them by adding the option of passing in a list of data fidelity dimensions. In the case of multiple data fidelity dimensions, a kernel is created for each dimension and added to a list, which is passed into a product kernel. This is useful for situations such as was discussed in #1942 with @Balandat.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/pytorch/botorch/blob/main/CONTRIBUTING.md#pull-requests)?

Yes.

## Test Plan

I changed code in `botorch/models/gp_regression_fidelity.py` and the accompanying test code in `test/models/test_gp_regression_fidelity.py`. The logic remains unchanged in the case of a single data-fidelity dimension. The changes for allowing multiple fidelity dimensions were tested by adding appropriate cases in the `FIDELITY_TEST_PAIRS` list in the `TestSingleTaskMultiFidelityGP` class.

Furthermore, the changes were tested on a minimisation toy problem with two data-fidelity dimensions and used to create the following [plots](https://github.com/AlexanderMouton/botorch/blob/ba3ca3e210615aa6818bdd04b832401bffd711bb/mdmfbo_toy_problem.pdf):

Screenshot from 2023-07-30 14-11-19

image

where g11 has fidelities [1.0, 1.0], g01 has fidelities [0.75, 1.0], g10 has fidelities [1.0, 0.75] and g00 has fidelities [0.75, 0.75].

## Related PRs

N/A

### CODE REVIEW

File: botorch/models/gp_regression_fidelity.py [hunk 1]
```diff
@@ -295,78 +298,114 @@ def _setup_multifidelity_covar_module(
...
-        }
+        subset_batch_dict = {}
+        for i in range(len(kernels)):
+            subset_batch_dict[f"{key_prefix}.{i}.raw_power"] = -2
+            subset_batch_dict[
+                f"{key_prefix}.{i}.covar_module_unbiased.raw_lengthscale"
+            ] = -3
+            subset_batch_dict[
+                f"{key_prefix}.{i}.covar_module_biased.raw_lengthscale"
+            ] = -3
```
Comment (Balandat) [inline botorch/models/gp_regression_fidelity.py:395]:
```suggestion
        subset_batch_dict = {}
        for i in range(len(kernels)):
            subset_batch_dict.update(
                {    
                    f"{key_prefix}.{i}.raw_power": -2,
                    f"{key_prefix}.{i}.covar_module_unbiased.raw_lengthscale": -3,
                    f"{key_prefix}.{i}.covar_module_biased.raw_lengthscale": -3,
                }
            }
```

**Review by Balandat — COMMENTED** [review]
Thanks for adding this and the thorough contribution. I have mostly cosmetic comments as well as a suggestion for how to accept data fidelities going forward (which may make fixing the lint errors unnecessary).

File: botorch/models/gp_regression_fidelity.py [hunk 2]
```diff
@@ -295,78 +298,114 @@ def _setup_multifidelity_covar_module(
         The covariance module and subset_batch_dict.
     """
 
-    if iteration_fidelity is not None and iteration_fidelity < 0:
-        iteration_fidelity = dim + iteration_fidelity
-    if data_fidelity is not None and data_fidelity < 0:
-        data_fidelity = dim + data_fidelity
+    if iteration_fidelity is not None:
+        if iteration_fidelity < 0:
+            iteration_fidelity = dim + iteration_fidelity
```
[the remark this answers is not in the document]
Comment (AlexanderMouton) [reply botorch/models/gp_regression_fidelity.py:303]: I mistakenly split this if statement and will revert the change, thanks!

File: botorch/models/gp_regression_fidelity.py (hunk 1 above)
Comment (AlexanderMouton) [reply botorch/models/gp_regression_fidelity.py]: Hi @Balandat 

Thank you for the feedback!
I've applied all recommended changes :)

**Comment by Balandat** [comment]
Thanks for making these edits. The only outstanding things would be to
1. rebase on the most recent changes in `main`, 
2. add a couple of tests to bring the coverage back up to 100% (via `assertWarnsRegex` and `assertRaisesRegex` to test the new deprecation / error setup)
3. update existing call sites in the tutorials, namely https://github.com/pytorch/botorch/blob/fe122b021c734860d2dd1c4a5bab9f821468ebdc/tutorials/multi_fidelity_bo.ipynb?short_path=aadaeba#L93 and https://github.com/pytorch/botorch/blob/fe122b021c734860d2dd1c4a5bab9f821468ebdc/tutorials/discrete_multi_fidelity_bo.ipynb?short_path=81261b8#L103

**Comment by AlexanderMouton** [comment]
Hi @Balandat 

I've restored the code coverage to 100%, made the changes to the tutorials' documentation and rebased to the newest changes. :)

**Comment by Balandat** [comment]
Hmm not sure why the doc build is failing, seems like there is an issue with the torch install? 
```
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.9.17/x64/lib/python3.9/site-packages/sphinx/ext/autodoc/importer.py", line 64, in import_module
    return importlib.import_module(modname)
  File "/opt/hostedtoolcache/Python/3.9.17/x64/lib/python3.9/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 680, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 850, in exec_module
  File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
  File "/opt/hostedtoolcache/Python/3.9.17/x64/lib/python3.9/site-packages/botorch/__init__.py", line 7, in <module>
    import gpytorch.settings as gp_settings
  File "/opt/hostedtoolcache/Python/3.9.17/x64/lib/python3.9/site-packages/gpytorch/__init__.py", line 5, in <module>
    import linear_operator
  File "/opt/hostedtoolcache/Python/3.9.17/x64/lib/python3.9/site-packages/linear_operator/__init__.py", line 2, in <module>
    from . import beta_features, operators, settings, utils
  File "/opt/hostedtoolcache/Python/3.9.17/x64/lib/python3.9/site-packages/linear_operator/beta_features.py", line 5, in <module>
    from .settings import _feature_flag
  File "/opt/hostedtoolcache/Python/3.9.17/x64/lib/python3.9/site-packages/linear_operator/settings.py", line 5, in <module>
    import torch
  File "/opt/hostedtoolcache/Python/3.9.17/x64/lib/python3.9/site-packages/torch/__init__.py", line 1157, in <module>
    from torch._C._VariableFunctions import *  # type: ignore[misc] # noqa: F[40](https://github.com/pytorch/botorch/actions/runs/5900934059/job/16014069040#step:6:41)3
ModuleNotFoundError: No module named 'torch._C._VariableFunctions'; 'torch._C' is not a package
```
I'll retry hopefully this is just a temporary issue with the package of the GHA runner.

**Comment by AlexanderMouton** [comment]
Hi @Balandat 

Is this an issue that I can fix from my side?

**Comment by Balandat** [comment]
This turned out to be an issue caused by a sphinx version bump and was fixed in #1990 - if you rebase on the latest changes in master the issue should go away.

**Comment by AlexanderMouton** [comment]
I've rebased to main again, so it should be fixed now @Balandat

**Pull request closed without merging**